cpp/include/cudf/detail/utilities/hash_functions.cuh

```diff
@@ -630,6 +628,34 @@ hash_value_type CUDA_DEVICE_CALLABLE SparkMurmurHash3_32<bool>::operator()(bool
   return this->compute<uint32_t>(key);
 }
 
+template <>
+hash_value_type CUDA_DEVICE_CALLABLE
+SparkMurmurHash3_32<int8_t>::operator()(int8_t const& key) const
```

Comment:
Why the arguments are `const&`? I just learn that passing a reference to a device callable function is dangerous---the original variable may reside in host memory.

Reply:
It's passed by reference because the hash function aliases the data as 32-bit integers to perform the hash on arbitrary types which may be expensive to pass by value.  We want the address of the inputs in memory to avoid that copy.  In this case all of the inputs are coming from values within a `cudf::column` and therefore cannot be in host memory.

Reply:
> I just learn that passing a reference to a device callable function is dangerous---the original variable may reside in host memory.

@ttnghia, this is only a problem when passing a variable by reference between device and host. In your case, that was a `column_device_view` constructed on `__host__` and passed to a `__device__` function. Note that the underlying pointer could have been copied with impunity.